src/main/java/io/cdap/plugin/gcp/bigquery/relational/BigQueryRelation.java

```diff
@@ -0,0 +1,103 @@
+package io.cdap.plugin.gcp.bigquery.relational;
+
+import io.cdap.cdap.etl.api.aggregation.DeduplicateAggregationDefinition;
+import io.cdap.cdap.etl.api.aggregation.GroupByAggregationDefinition;
+import io.cdap.cdap.etl.api.relational.Expression;
+import io.cdap.cdap.etl.api.relational.Relation;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * An implementation of {@link Relation} designed to operate on BigQuery.
+ */
+public class BigQueryRelation implements Relation {
+  private final boolean isValid;
+  private final String validationError;
+
+  private final Set<String> columns;
+  private final Map<String, Expression> columnValues;
+  private final Expression filter;
+
+  /**
+   * Initializes a new {@link BigQueryRelation} using the set of columns specified.
+   * @param columns
+   */
+  public BigQueryRelation(Set<String> columns) {
+    this(columns, new HashMap<>(), null);
+  }
+
+  private BigQueryRelation(Set<String> columns,
+                           Map<String, Expression> columnValues,
+                           Expression filter) {
+    this(columns, columnValues, filter, true, null);
+  }
+
+  private BigQueryRelation(Set<String> columns,
+                           Map<String, Expression> columnValues,
+                           Expression filter,
+                           boolean isValid,
+                           String validationError) {
+    this.columns = columns;
+    this.columnValues = columnValues;
+    this.filter = filter;
+    this.isValid = isValid;
+    this.validationError = validationError;
+  }
+
+  private Relation getInvalidRelation(String validationError) {
+    return new BigQueryRelation(null, null, null, false, validationError);
+  }
+
+  @Override
+  public boolean isValid() {
+    return isValid;
+  }
+
+  @Override
+  public String getValidationError() {
+    return isValid() ? null : validationError;
+  }
+
+  @Override
+  public Relation setColumn(String column, Expression value) {
+    Set<String> newColumns = new HashSet<>(columns);
+    Map<String, Expression> newColumnValues = new HashMap<>(columnValues);
+    newColumns.add(column);
+    newColumnValues.put(column, value);
+
+    return new BigQueryRelation(newColumns, newColumnValues, filter);
```

Comment:
Passing filter like this won't work as filter may depend on other columns being dropped. 
E.g. `r.setFilter("a == 5").dropColumn("a")` won't work

Reply:
Actually, same applies to any operations, e.g. consider `r.setColumn("a", "a+b").setColumn("a", "a/c")`. You should remember all nesting. You may do flattening, but only if you ensure correctness

Reply:
`r.setFilter("a == 5").dropColumn("a")` is a valid statement, though.

Input: columns a,b,c
`r.setFilter("a = 5").dropColumn("a")` means:

`select b,c from <table> where a = 5`

I do agree with `r.setColumn("a", "a+b").setColumn("a", "a/c")` is invalid, but I'm inclined to say this should simply override the initial definition with the latest one.

Reply:
`r.setColumn("a", "a+b").setColumn("a", "a/c")` is actually valid as you have a new Relation created after each call. Each time you do an op you do on top of existing relation.
In SQL is expressed as subselect, e.g. `select <columns except a>, a/c as a from (select <columns except a>, a+b as a  from table)`.
Overall quite often this can be optimized to remove nested subselects, but I am not sure if we need to do it. We can just go with subselects now, DB optimizers can often handle it. One special case can be adding a new column - in this case subselect can be omited. This would help to optimize r.setColumn("a1", expr).setColumn("a2", expr).... use case

Reply:
But point is that when you have a relation it should work irrespective of if it's a table or derived relation.
So after
`Relation x = r.setColumn("a", "a+b")` you have a relation x with column a set to certain values that is different from values in relation r. And `x.setColumn("a", "a/c")` should give correct result different from `r.setColumn("a", "a/c")`

Reply:
I see, I thought the pattern was more similar to a Builder.

The only potential red flag with nested statements is that there may be limits in how many "layers" of nesting the underlying engine may support. This can cause issues.

Another potential solution is to tokenize the expression and replace each token with the latest "expression" for this token.

Reply:
So `rel.setColumn("a","a+b").setColumn("a","a/c") becomes something like

`Select ... , (a+b)/c as a ...`

Reply:
Yep, relation is immutable. You can use a single relation and produce multiple derivative relations (e.g. to join). Generally it's very similar to spark DataSet  concept. It's not a builder as there is no "build" to mark "I am done with this" and introducing build would be unnecessary API complications

Reply:
I tried to highlight it in the method javadocs (e.g. https://github.com/cdapio/cdap/blob/6c60f2d8ebeb871690382a10caf40aa532a384f7/cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/relational/Relation.java#L72), but may be it should be made more clear in the class javadoc